fix: Codec for article slugs

pete-murphy opened the pull request:
Allow for non-alphanumeric and non-lower-case characters in slugs.

Fixes #127.

petermurphy-welldot committed:
fix: Codec for article slugs

Allow for non-alphanumeric and non-lower-case characters in slugs.

Fixes #127.

pete-murphy commented on src/Data/Slug.purs:
```diff
@@ -0,0 +1,120 @@
+import Data.String.CodePoints (codePointFromChar, fromCodePointArray, toCodePointArray)
+import Data.String.Pattern (Pattern(..), Replacement(..))
+
+-- | A `Slug` represents a string value which is guaranteed to have the
+-- | following qualities:
+-- |
+-- | - it is not empty
+-- | - it consists of groups of characters separated by `-` dashes,
+-- |   where the slug cannot begin or end with a dash, and there can
+-- |   never be two dashes in a row.
+-- |
+-- | Example: `Slug "this-is-an-article-slug"`
+newtype Slug = Slug String
```
This was copied from https://github.com/thomashoneyman/purescript-slug/blob/0142716b49758829d5b48244fe1e9f87ea935b77/src/Slug.purs#L21-L31, removing the following guarantees: "every character with a defined notion of case is lower-cased" and "it consists of _alphanumeric_ groups of characters".

pete-murphy commented on src/Data/Slug.purs:
```diff
@@ -0,0 +1,120 @@
+  -- Replace non-Latin 1 characters with spaces to be stripped later.
+  onlyAlphaNum =
+    fromCodePointArray
+      <<< map (\x -> if isLatin1 x then x else codePointFromChar ' ')
```
Copied from https://github.com/thomashoneyman/purescript-slug/blob/0142716b49758829d5b48244fe1e9f87ea935b77/src/Slug.purs#L75, removing `isAlphaNum x`

pete-murphy commented on src/Data/Slug.purs:
````diff
@@ -0,0 +1,120 @@
+-- | ```
+generate :: String -> Maybe Slug
+generate s = do
+  let arr = words $ onlyAlphaNum $ stripApostrophes s
````
Copied from https://github.com/thomashoneyman/purescript-slug/blob/0142716b49758829d5b48244fe1e9f87ea935b77/src/Slug.purs#L63, removing `String.toLower`

pete-murphy commented on src/Data/Slug.purs:
````diff
@@ -0,0 +1,120 @@
+-- | ```
+generate :: String -> Maybe Slug
+generate s = do
+  let arr = words $ onlyLatin1 $ stripApostrophes s
````
Copied from https://github.com/thomashoneyman/purescript-slug/blob/0142716b49758829d5b48244fe1e9f87ea935b77/src/Slug.purs#L63 removing `String.toLower`

thomashoneyman commented on src/Data/Slug.purs:
```diff
@@ -0,0 +1,120 @@
+derive newtype instance Semigroup Slug
+
+instance Show Slug where
+  show (Slug str) = "(Slug " <> str <> ")"
```
I think we should ditch these instances since they aren’t used here (I generally don’t use Show, and we use codecs instead of JSON instances)

pete-murphy replied:
I removed the `*Json`, `Show`, and `Semigroup` instances (leaving `Eq` and `Ord` since those are used in `instance ordRoute :: Ord Route`).

thomashoneyman commented on src/Data/Slug.purs:
```diff
@@ -0,0 +1,120 @@
+truncate n (Slug s)
+  | n < 1 = Nothing
+  | n >= String.length s = Just (Slug s)
+  | otherwise = generate $ String.take n s
```
```suggestion
  | otherwise = generate $ String.take n s

```

thomashoneyman commented:
Thanks! I’m fine integrating this module into the code base to fix this issue. Looks like we should also remove the import from Spago?

pete-murphy commented:
> Looks like we should also remove the import from Spago?

Will do 👍  I'm surprised there's no warning from `spago build` about unused dependencies, maybe that's not a feature in new `spago` yet?

pete-murphy commented:
Also, I see
```
Lockfile is out of date, generating it...
Lockfile written to spago.lock. Please commit this file.
```
when running `spago build`, but `*.lock` in `.gitignore` ignores it, should there be an exception for this `spago.lock` file?

**Edit:** I added it here [ced03b5](https://github.com/thomashoneyman/purescript-halogen-realworld/pull/128/commits/ced03b5d2ddc0fc32fdf072f585ddcf543ca18e7) but I can remove that commit if we don't want to track `spago.lock` in Git.

pete-murphy committed:
fix: PR feedback

Removing unused `Semigroup`, `Show`, `EncodeJson`, and `DecodeJson`
instances from `Slug` module, adding newline at end of file, and
removing unused `slug` dependency from `spago.yaml`.

pete-murphy committed:
chore: Commit spago.lock

pete-murphy commented:
> I’m fine integrating this module into the code base to fix this issue.

I created an issue in `slug` in case there is interest in an upstream fix https://github.com/thomashoneyman/purescript-slug/issues/3

thomashoneyman commented:
We can use the —pedantic-packages flag to get warnings for unused dependencies. And yes, the lock being ignored is no longer relevant, we do want the spago lock committed!

thomashoneyman commented:
I merged this because it fixes the existing issue — but if you have follow up things you’d like to add I’m all ears!
